content/blog/2022/2022-02-14-spring-shell.md

```diff
@@ -0,0 +1,331 @@
+
+
+## Dynamic Command Availability
+Some commands only make sense when certain pre-conditions are met. For example, a `sign-out` command should be available only if a `sign-in` command has been issued, and if the user tries to run the `sign-out` command we want to warn them that it's not possible.
+
+Spring Shell offers us **three** ways to achieve our goal.
+
+### The First Way
+It checks our class for a method with a special name and with a return type of `Availability`.  
+The special name has to be in the format `command-to-checkAvailability`:
+```java  
+@ShellComponent 
+public class SSHLoggingCommand {    
+    Logger log = Logger.getLogger(SSHLoggingCommand.class.getName());    
+    private boolean signedIn;    
+    
+    @ShellMethod(value = "sign in")    
+    public void signIn()    
+    {    
+        this.signedIn = true;    
+        log.info("Signed In!");    
+    }    
+    
+    @ShellMethod(value = "sign out")    
+    public void signOut()    
+    {    
+        this.signedIn = false;    
+        log.info("Signed out!");    
+    }    
+    // note the naming     
+   public Availability signOutAvailability()    
+    {    
+        return signedIn ?    
+                Availability.available() : Availability.unavailable("Must be signed in first");    
+	} 
+}  
+``` 
+So if we try to run the `sign-out` command without first signing in we will get the following message:
+```bash  
+shell:>sign-out  
+Command 'sign-out' exists but is not currently available because Must be signed in firstDetails of the error have been omitted. You can use the stacktrace command to print the full stacktrace.shell:>  
+```  
+
+### The Second Way
+Uses the `@ShellMethodAvailability` annotation, in which we specify the method name we want to use to Availability check:
+```java  
+@ShellComponent  
+public  class  SSHLoggingCommand  
+{
+	...
+	@ShellMethod(value = "sign out") 
+	@ShellMethodAvailability("signOutCheck") 
+	public void signOut() 
+	{    
+	    this.signedIn = false;    
+		log.info("Signed out!"); 
+	}  
+		
+	public Availability signOutCheck() 
+	{    
+	   return signedIn ?  Availability.available() : Availability.unavailable("Must be signed in first"); 
+	}  
+}
+``` 
+
+### The Third Way
+It enables us to have several methods attached to a single availability method.  
+We are going to use the annotation `ShellMethodAvailability` with an array of the **commands names (not method names)**:
+
+```java  
+@ShellComponent  
+public  class  SSHLoggingCommand  
+{
+	...
+	@ShellMethod(value = "sign out") public void signOut() 
+	{    
+	    this.signedIn = false;    
+		log.info("Signed out!"); 
+	}    
+
+	 @ShellMethod(value = "Change password") 
+	 public void changePass(@ShellOption String newPass) 
+	 {    
+		log.info(format("Changed password to '%s'", newPass)); 
+	 }    
+	 
+	 @ShellMethodAvailability({"sign-out", "change-pass"}) 
```

Comment:
Here it's `sign-out`, but the method is annotated with `sign out`. Which is it?

Reply:
Yeah in this case the resulting command name will be `sign-out` and not `sign out`, which is the description of the command. The annotation `ShellMethodAvailability` requires the resulting command name.